Support arbitrary axes for ReinterpretArray

timholy opened the pull request:
This allows one to reinterpret arrays where indexing does not start at 1. Axis 1 of the reinterpreted array starts at the same index as the parent. Consequently, if the parent axis 1 is `f:l`, then the reinterpreted array has axis `f:(l-f+1)/r+f-1`, where `r` is the ratio of sizes of the eltypes.

The need for this was discovered while porting ImageCore, which used to export its own types that performed this operation for one special case, converting between numbers and colors.

vtjnash commented:
I was wondering that too. I suppose the counter-point is that the contraction point for a normal array is at `1` – e.g. it's a dynamic resizing of the length offset from the first value, not about the origin.

Keno commented:
> I was wondering that too. I suppose the counter-point is that the contraction point for a normal array is at 1 – e.g. it's a dynamic resizing of the length offset from the first value, not about the origin.

Or it's about the origin with rounding away from 0.

Keno commented:
I'm not sure I've fully understood what you're saying, but wouldn't exchanging the
reshape and the reinterpret do what you want here?
```
reinterpret(N0f8, reshape(imgo, 1:1, axes(imgo)...))
```

timholy commented:
Yes. Maybe to say it more clearly, what I'm proposing is that instead of doing something magical to the first dimension, `reinterpret` may change the dimensionality. Specifically, for a call `reinterpret(T, A::AbstractArray{S})`:

- if `sizeof(T) < sizeof(S)`, `reinterpret` adds a new first dimension to `A` of length `sizeof(S)  ÷ sizeof(T)`;
- if `sizeof(T) < sizeof(S)`, `reinterpret` "eats" the first dimension of `A`, which *must* be of length `sizeof(T) ÷ sizeof(S)`;
- if `sizeof(T) == sizeof(S)`, the dimensionality and shape of `A` are unaltered.

The advantage of this proposal is that there's essentially nothing arbitrary to what `reinterpret` does to the axes of the array.

Keno commented:
> The advantage of this proposal is that there's essentially nothing arbitrary to what reinterpret does to the axes of the array.

Yes, I can see the appeal. I think I would be ok with this proposal, though I think people may not like having to reshape first to e.g. `reinterpret(UInt64, A::Vector{UInt8})`.

timholy commented:
Yes, agreed.

I guess one other way we could do this would be to insist that the first dimension start indexing with 1. Any later dimensions could be arbitrary.

timholy committed:
Support offset axes in reinterpret

The first axis of the input must start with 1. The remaining axes
can be arbitrary.

timholy commented:
I reworked this to insist that axis1 starts at 1. By and large this patch is now just "implement `axes(::ReinterpretArray)`".

mbauman commented:
Ok, so the rule this now implements is that we only support reinterpret for offset arrays if the two sizes are identical — in that case `axes(R) == axes(A)` (where `R = reinterpret(T, A)`).  If the sizes are different, the reinterpret fails.  If `A`'s first axis is one-indexed, then we support reinterpret, with `axes(R)[2:end] == axes(A)[2:end]` and `axes(R, 1)` being `1:N`.

This seems reasonable, assuming I got that correct.

timholy commented:
Right, although I might phrase it this way:
1. if `sizeof(T) == sizeof(S)` then `reinterpret` always succeeds (and `axes(R) == axes(A)`)
2. otherwise,
    a. we require that `length(axes(A, 1)*sizeof(S)` is an integer multiple of `sizeof(T)`
    b. `axes(A, 1)` starts with 1

Only 2b is new here. (And it's not really new, it's just that we don't fail via the use of `size`.)

timholy commented:
Test failures seem unrelated. Good to go?

timholy commented:
I'll merge this by tonight so I have a fighting chance of getting ImageCore working on 0.7 before I disappear for a couple weeks. That will hopefully let other people get started on the rest of the Images world (which is proving to be a useful stress-test for 0.7).

StefanKarpinski commented:
Circle CI failure is httpbin flaking. There's a doctest failure on Travis Linux 64, which appears to be https://github.com/JuliaLang/julia/pull/28015—why would that only be failing on that platform?

mbauman commented:
We only run doctests on travis 64-bit linux:

https://github.com/JuliaLang/julia/blob/2b086642be9dc89153991f844d9be83ec6c26d5c/.travis.yml#L148-L150

StefanKarpinski commented:
Ok, so this should be safe to merge than since that PR fixed the doc test?
